maint/qmpi_binding.py

```diff
@@ -265,38 +266,84 @@ def generate_code(parseset: Mapping[str, Any], postfix: str, kind_map: Mapping[s
 enum_list_string = (',\n    '.join(enum_list));
 qmpi_h_file.write(f"enum QMPI_Functions_enum {{" "\n    "
                   f"{enum_list_string}," "\n"
+                   "MPI_FILE_C2F_T,\n"
+                   "MPI_FILE_F2C_T,\n"
                    "MPI_LAST_FUNC_T\n"
                   f"}};" "\n\n");
 
 qmpi_h_file.write(""
         "typedef struct {\n"
-        "    void *storage_stack;\n"
+        "    void **storage_stack;\n"
+        "    void *calling_address;\n"
         "} QMPI_Context;\n"
         "\n"
         );
 
 qmpi_h_file.write(''.join(qmpi_h_file_text));
 
+qmpi_h_file.write(
+    "MPI_Fint QMPI_File_c2f(QMPI_Context context, int tool_id, MPI_File fh) MPICH_API_PUBLIC;\n"
+    "typedef MPI_Fint (QMPI_File_c2f_t) (QMPI_Context context, int tool_id, MPI_File fh);\n"
+    "\n"
+    "MPI_File QMPI_File_f2c(QMPI_Context context, int tool_id, MPI_Fint fh) MPICH_API_PUBLIC;\n"
+    "typedef MPI_File (QMPI_File_f2c_t) (QMPI_Context context, int tool_id, MPI_Fint fh);\n"
+    "\n"
+);
+
 # Add the types and registration functions to the header file
 qmpi_h_file.write("#define QMPI_MAX_TOOL_NAME_LENGTH 256\n\n");
 qmpi_h_file.write("int QMPI_Register_tool_name(const char *tool_name, void (* init_function_ptr)(int tool_id)) MPICH_API_PUBLIC;\n");
 qmpi_h_file.write("int QMPI_Register_tool_storage(int tool_id, void *tool_storage) MPICH_API_PUBLIC;\n");
 qmpi_h_file.write("int QMPI_Register_function(int calling_tool_id, enum QMPI_Functions_enum function_enum,\n"
                   "                           void (* function_ptr)(void)) MPICH_API_PUBLIC;\n");
 qmpi_h_file.write("\n");
-qmpi_h_file.write("#include <stddef.h>\n"
+qmpi_h_file.write("#include <execinfo.h>\n#include <stddef.h>\n#include <stdio.h>\n#include <stdlib.h>\n#include <string.h>\n"
         "extern MPICH_API_PUBLIC void (**MPIR_QMPI_pointers) (void);\n"
         "extern MPICH_API_PUBLIC void **MPIR_QMPI_storage;\n"
+        "#define BACKTRACE_SIZE 3\n"
+        "#define BACKTRACE_ENTRY 1\n"
         "\n"
-        "static inline void QMPI_Get_function(int calling_tool_id, enum QMPI_Functions_enum function_enum,\n"
+        "static inline int QMPI_Get_function(int calling_tool_id, enum QMPI_Functions_enum function_enum,\n"
         "                      void (**function_ptr) (void), QMPI_Context *next_tool_context,\n"
         "                      int *next_tool_id)\n"
         "{\n"
         "    int mpi_errno = MPI_SUCCESS;\n"
         "    QMPI_Context context;\n"
         "\n"
+        "    void *array[BACKTRACE_SIZE];\n"
+        "    char **traces;\n"
+        "    int size, i;\n"
+        "    char *target = NULL;\n"
+        "\n"
         "    context.storage_stack = MPIR_QMPI_storage;\n"
         "\n"
+        "    // Capture the calling function address\n"
```

Comment:
We don't use C++ style comments in MPICH. Can you change these to `/* */`?

Reply:
Thanks, will fix it.